### PR TITLE
API: submenus extension point

### LINKED ISSUE #100172: Add API support for submenus
Add support for submenus to simplify the menu structure of the Git extensions '...' menu.

### PR DESCRIPTION
**Pull request by joaomoreno**
This PR fixes #100172

- [x] Fix non-editor submenus, fix dropdown submenus: address `SubmenuItemAction` vs `ContextSubMenu` confusion
- [x] Fix native context menus
- [x] Support nested submenus
- [x] Prevent submenu cycles
- [x] Expand the set of menus which have submenu support
- [x] Support top-level toolbar submenus, including icon for the action item

### CODE REVIEW

**Review by jrieken — APPROVED** [review]

**Comment by joaomoreno** [comment]
@sbatten @isidorn Native submenus were yet another implementation, dependent on the bogus `ContextSubMenu` class. I deleted that class.

**Comment by joaomoreno** [comment]
@sbatten @jrieken Recursive submenus work in custom UI:

image

But they break down in native UI, since the entire menu hierarchy must be given to Electron at the time of creation:

image

---

Because of the limitation and the fact that it would simply be a gimmick, I'll write some code to prevent menu cycles.

**Review by sbatten — APPROVED** [review]

**Comment by joaomoreno** [comment]
@jrieken I've refactored `menusExtensionPoint.ts` heavily to make it easier to maintain the list of API menus, give it a look: https://github.com/microsoft/vscode/pull/102784/commits/c3cef6f8ee46ff7a4fa664d63c656ba96dfc1313

**Comment by joaomoreno** [comment]
Submenus now work across most menus! Check this out:

Annotation 2020-07-23 152725
Annotation 2020-07-23 153320
Annotation 2020-07-23 153406

**Comment by isidorn** [comment]
Very cool work 👏

File: src/vs/base/browser/ui/dropdown/dropdownActionViewItem.ts [hunk 1]
```diff
@@ -57,11 +58,17 @@ export class DropdownMenuActionViewItem extends BaseActionViewItem {
...
+			this.element = append(el, $('a.action-label'));
+
+			const classNames = this.options.classNames ? asArray(this.options.classNames) : [];
+
+			// todo@aeschli: remove codicon, should come through `this.options.classNames`
+			if (!classNames.find(c => c === 'icon')) {
+				classNames.push('codicon');
 			}
 
+			addClasses(this.element, ...classNames);
```
Comment (jrieken) [inline src/vs/base/browser/ui/dropdown/dropdownActionViewItem.ts:70]: fyi - es6 deprecated, use `this.element.classList...`

File: src/vs/workbench/api/common/menusExtensionPoint.ts [hunk 2]
```diff
@@ -47,16 +47,15 @@ namespace schema {
 			case 'explorer/context': return MenuId.ExplorerContext;
 			case 'editor/title/context': return MenuId.EditorTitleContext;
 			case 'debug/callstack/context': return MenuId.DebugCallStackContext;
-			case 'debug/toolbar': return MenuId.DebugToolBar;
 			case 'debug/toolBar': return MenuId.DebugToolBar;
```
Comment (jrieken) [inline src/vs/workbench/api/common/menusExtensionPoint.ts]: API breakage?

Comment (isidorn) [reply src/vs/workbench/api/common/menusExtensionPoint.ts]: Yeah we had both for backwards compatibility, even though debug/toolbar was "deprecated". I am not sure if there are extensions which really on this.

Comment (jrieken) [reply src/vs/workbench/api/common/menusExtensionPoint.ts]: High risk. You could know relatively easy (telemetry, search extension usages) or just keep it given the low price of supporting it. btw: deprecated doesn't mean that we will remove it

Comment (isidorn) [reply src/vs/workbench/api/common/menusExtensionPoint.ts]: I just searched top 1000 extensions usages on `debug/toolbar` usage and only one extension used the deprecated name - the extension is not that popular and I have created an issue for then.
So I suggest the following: let's go with the removal, and if needed we can easily reintroduce it as part of a recovery build if somebody complains.
fyi @weinand

Comment (joaomoreno) [reply src/vs/workbench/api/common/menusExtensionPoint.ts]: Great catch @jrieken, I really thought this was an oversight and failed to see the case difference. I'll leave it up to @isidorn and @weinand. Sorry all!

File: src/vs/base/browser/ui/dropdown/dropdownActionViewItem.ts (hunk 1 above)
Comment (joaomoreno) [reply src/vs/base/browser/ui/dropdown/dropdownActionViewItem.ts:70]: Actually didn't know that!

**Comment by joaomoreno** [comment]
@jrieken Thanks for the post-mergeum review! 🙏

**Pull request merged**